PWA shortcuts how-to page (#27119)
mdn/content: files/en-us/web/progressive_web_apps/how_to/expose_common_actions_as_shortcuts/index.md

---

> title: Expose common app actions as shortcuts

Reviewer: ```suggestion
title: PWA shortcut action menu
```
I am not sure if this is the best option either; alternatively, `Creating shortcut action menus` but since it's in the "how to" maybe just `Shortcut action menus`. I realize the PWA is redundant since we're in the PWA section, but for search results, it will be helpful. Basically, looking for something that people will instantly understand and may actually enter into a search engine.

Reviewer 2: I think the grammar of these should work like "How to / Do the thing", so the title should be verb-y. This is the pattern for all the existing How-tos.

I'm also -1 on including "PWA" in the title, or at least, I'd like to understand better when we should and when we should not do this.

Author: Yeah, I'd prefer to stick to the pattern we already have.
Let's have a discussion about using PWA or not, and then change all articles in one go to match.

Reviewer 2: > I'm also -1 on including "PWA" in the title

Actually I take this back - the other how-tos have "app" and the one I just filed uses "PWA" in the title. So perhaps we should choose to use "app" or "PWA" consistently, but so far we have tended to include one or the other in how-to titles.

slug: Web/Progressive_web_apps/How_to/Expose_common_actions_as_shortcuts
---

{{PWASidebar}}

> Many operating systems support showing shortcut menus when the user right-clicks or long-presses an app icon. For example, on Windows, right-clicking on any pinned program in the taskbar shows a list of program-specific actions and recently opened files:

Reviewer: ```suggestion
Many operating systems support showing shortcut menus, or jump-list, when the user right-clicks or long-presses an app icon. For example, on Windows, right-clicking on any pinned program in the taskbar shows a list of program-specific actions and recently opened files:
```

![The task bar in Windows, showing several pinned apps. The Firefox app icon was right-clicked, and the jump list is displayed, showing frequent tabs and common tasks](./jump-list.png)

On Android, long-pressing an app icon also shows a list of common app actions:

![The Android app launcher, showing an app icon that's been long-pressed. The shortcut menu is displayed, showing common actions](./android-shortcuts.png)

[Progressive Web Apps (PWAs)](/en-US/docs/Web/Progressive_web_apps) can be installed on devices just like platform-native apps and, like their native counterparts, they can also define app shortcut menus to let users access common actions.

> Shortcuts are only displayed by right-clicking or long-pressing the app icon, which means that they're only available once the PWA is installed on the user's device. To learn how to make your PWA installable, see [Making PWAs installable](/en-US/docs/Web/Progressive_web_apps/Guides/Making_PWAs_installable).

Reviewer: ```suggestion
Shortcuts are only displayed by right-clicking or long-pressing the app icon, meaning they're only available once the PWA is installed on the user's device. To learn how to make your PWA installable, see [making PWAs installable](/en-US/docs/Web/Progressive_web_apps/Guides/Making_PWAs_installable).
```

## Why use shortcuts?

Defining shortcuts for your PWA can make users more productive by letting them access your app's main actions directly from their home screen. Additionally, defining shortcuts can help make your PWA feel more platform-native and therefore more familiar to your users.

## Define shortcuts in the web app manifest

> To define shortcuts for your PWA, use the [`shortcuts`](/en-US/docs/Web/Manifest/shortcuts) member of the web app manifest. This member is an array of objects, each of which defines the shortcut name, URL, and optional short name, description, and icons. For example, here's the web app manifest of a calendar app that defines two shortcuts:

Reviewer: ```suggestion
To define shortcuts for your PWA, use the [`shortcuts`](/en-US/docs/Web/Manifest/shortcuts) member of the web app manifest. This member is an array of objects defining each shortcut's name and URL, as well as the optional short name, description, and icons. For example, here's the web app manifest of a calendar app that defines two shortcuts:
```
We likely should link the web app manifest to the intro-level tutorial on how to write a manifest file once that is pushed. Or, do you have a better link?

Author: In the meantime, we can link to the web app manifest ref page I guess: https://developer.mozilla.org/en-US/docs/Web/Manifest

```json
{
  "name": "Calendar",
  "start_url": "/",
  "display": "standalone",
  "icons": [
    {
      "src": "images/icon-256.png",
      "sizes": "256x256",
      "type": "image/png"
    }
  ],
  "shortcuts": [
    {
      "name": "New event",
      "url": "/new-event"
    },
    {
      "name": "View today's events",
      "url": "/today"
    }
  ]
}
```

The most important properties of each shortcut object are:

- `name`

>   - : The name of the shortcut, which is displayed in the shortcut menu. MAke sure to keep it short but also descriptive enough to let users know what the shortcut does.

Reviewer 2: ```suggestion
  - : The name of the shortcut, which is displayed in the shortcut menu. Make sure to keep it short but also descriptive enough to let users know what the shortcut does.
```

- `url`
  - : The URL to launch the PWA with when the user selects the shortcut. This URL can be absolute, in which case it should exist within the [scope](/en-US/docs/Web/Manifest/scope) of the web app manifest. The URL can also be relative, in which case it's resolved relative to the PWA's [start URL](/en-US/docs/Web/Manifest/start_url).

All other shortcut object properties are optional, but you should consider providing them to make the shortcut more useful to users:

- `short_name`
  - : A short name for the shortcut, which is displayed when there's not enough space to display the full name.
- `description`
  - : A description of the shortcut. This string can be accessed by assistive technologies, such as screen readers, to help users understand what the shortcut does.
- `icons`

>   - : An array of image objects to display in the shortcut menu. Each image object is processed just like the [`icons`](/en-US/docs/Web/Manifest/icons) member of the web app manifest, and can be used to provide different-sized icons for different device requirements.

Reviewer: would be helpful to explain the size requirement for these icons or link to a good resource. Alternatively, we can add and icon to the example

Author: I haven't really found the right docs for this yet. Google has some docs on web.dev here: https://web.dev/app-shortcuts/#define-app-shortcuts-in-the-web-app-manifest

> If you want pixel-perfect icons, provide them in increments of 48dp (i.e. 36x36, 48x48, 72x72, 96x96, 144x144, 192x192 pixel icons). Otherwise, it is recommended that you use a single 192x192 pixel icon.
> As a quality measure, icons must be at least half of the device's ideal size on Android, which is 48dp. For example, to display on an [xxhdpi screen](https://developer.android.com/training/multiscreen/screendensities#TaskProvideAltBmp), the icon must be at least 72 by 72 pixels. (This is derived from the [formula for converting](https://developer.android.com/training/multiscreen/screendensities#dips-pels) dp units for pixel units.)

But this is likely Android-specific only, so I thought I'd rather not say anything, than something I'm not sure about.
